File: app/src/main/java/com/estasvegano/android/estasvegano/model/ProducerModel.java
package com.estasvegano.android.estasvegano.model;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import com.estasvegano.android.estasvegano.data.web.EVeganoApi;
import com.estasvegano.android.estasvegano.data.web.ErrorParser;
import com.estasvegano.android.estasvegano.data.web.request.AddProducerRequest;
import com.estasvegano.android.estasvegano.entity.Producer;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.List;

import rx.Observable;

public class ProducerModel {

    @NonNull
    private final EVeganoApi api;

    @NonNull
    private final ErrorParser errorParser;

    @NonNull
    private final ObjectMapper objectMapper;

    public ProducerModel(
            @NonNull EVeganoApi api,
            @NonNull ErrorParser errorParser,
            @NonNull ObjectMapper objectMapper) {
        this.api = api;
        this.errorParser = errorParser;
        this.objectMapper = objectMapper;
    }

    @CheckResult
    @NonNull
    public rx.Single<Producer> addProducer(@NonNull String title) {
        return api.addProducer(AddProducerRequest.builder().title(title).build())
                .flatMap(errorParser::checkIfError)
                .map(data -> objectMapper.convertValue(data, Producer.class));
    }

    @CheckResult
    @NonNull
    public rx.Single<List<Producer>> getProducers(@NonNull String title) {
        //noinspection unchecked
        return api.getProducers(title)
                .flatMap(errorParser::checkIfError)
                .flatMapObservable(Observable::from)
                .map(data -> objectMapper.convertValue(data, Producer.class))
                .toList()
                .toSingle();
    }
}

File: app/src/main/java/com/estasvegano/android/estasvegano/model/ProducerModel.kt
package com.estasvegano.android.estasvegano.model

import android.support.annotation.CheckResult
import com.estasvegano.android.estasvegano.data.web.EVeganoApi
import com.estasvegano.android.estasvegano.data.web.ErrorParser
import com.estasvegano.android.estasvegano.data.web.request.AddProducerRequest
import com.estasvegano.android.estasvegano.entity.Producer
import com.fasterxml.jackson.databind.ObjectMapper
import io.reactivex.Single

class ProducerModel(
        private val api: EVeganoApi,
        private val errorParser: ErrorParser,
        private val objectMapper: ObjectMapper
) {

    @CheckResult
    fun addProducer(title: String): Single<Producer> {
        return api.addProducer(AddProducerRequest(title))
                .flatMap { errorParser.checkIfError(it) }
                .map<Producer> { data -> objectMapper.convertValue<Producer>(data, Producer::class.java) }
    }

    @CheckResult
    fun getProducers(title: String): Single<List<Producer>> {
        return api.getProducers(title)
                .flatMap { errorParser.checkIfError(it) }
                .map { (it as Iterable<Any>).map { objectMapper.convertValue<Producer>(it, Producer::class.java) } }
    }
}
